### DISCUSSION TITLE
How to get Signature Keyboard and touchscreen working on Surface Pro 8

### DISCUSSION THREAD

**Discussion by mickpf** [linux-surface/linux-surface#981] [Q&A] [answered]
Hello,

I try to get Ubuntu 22.04 with the linux-surface kernel patches for 5.15.x running on the Surface Pro 8.

Touchscreen, the Signature Keyboard and the touch panel do not work as without the patches.

Which drivers/modules must be included/loaded to get them working?

Kind Regards,
Michael

**Comment by qzed** [comment, marked as answer]
Please use a recent set of patches, i.e. the ones from `patches/6.0` or one of our recent v6.0.x pre-built kernels. The older patches are just archived versions, therefore outdated.

That being said, the keyboard should work with the v5.15 patches. Did you make sure that you have all required config options set from `configs/surface-5.15.config`?

Touch will however not work with those patches. At the moment, you'll need to install an external driver for that. See https://github.com/quo/ithc-linux/ (and you'll also need to install https://github.com/linux-surface/iptsd from the repo directly).

**Comment by mickpf** [comment]
Many thanks for your answer! I'll try to apply it next days.

**Comment by mickpf** [comment]
I've followed your instruction completely with the kernel source 6.0.7...
Loading the module "ithc" generates dmesg output
```
[  447.828617] ithc: loading out-of-tree module taints kernel.
[  447.829737] ithc 0000:00:10.6: enabling device (0000 -> 0002)
[  448.049039] ithc 0000:00:10.6: config: e0000002 00000000 001fc0ff 00000010 00000014 43495424 09e34de8 0c37045e 00000001 e6eb47ab 1ae6d4ad 00000000 6e155668 00040302 3aad6885 c3ee4710
[  448.049866] ithc 0000:00:10.6: registered device ithc
[  448.050993] DMAR: DRHD: handling fault status reg 2
[  448.051003] DMAR: [INTR-REMAP] Request device [01:05.0] fault index 0x40 [fault reason 0x26] Blocked an interrupt request due to source-id verification failure
[  449.098841] ithc 0000:00:10.6: ithc_probe: hid_add_device failed with -110
[  449.098856] ithc: probe of 0000:00:10.6 failed with error -110
```
I've configured almost relevant kernel drivers as modules. Maybe too much drivers...
```
8250                   49152  1 8250_dw
8250_base              53248  2 8250,8250_dw
8250_dw                24576  0
ac97_bus               16384  1 snd_soc_core
acpi_pad               20480  0
aesni_intel           393216  9
af_alg                 28672  6 algif_hash,algif_skcipher
algif_hash             16384  1
algif_skcipher         16384  1
atomisp_gmin_platform    28672  0
binfmt_misc            16384  1
bluetooth             634880  44 btrtl,btmtk,btintel,btbcm,bnep,btusb,rfcomm
bnep                   20480  2
btbcm                  24576  1 btusb
btintel                36864  1 btusb
btmtk                  16384  1 btusb
btrtl                  24576  1 btusb
btusb                  57344  0
ccm                    20480  6
cfg80211              430080  3 iwlmvm,iwlwifi,mac80211
clk_tps68470           16384  0
configfs               40960  1
coretemp               16384  0
crc32_pclmul           16384  0
cryptd                 20480  3 crypto_simd,ghash_clmulni_intel
crypto_simd            16384  1 aesni_intel
ecc                    36864  1 ecdh_generic
ecdh_generic           16384  2 bluetooth
ghash_clmulni_intel    16384  0
gpio_keys              20480  0
i2c_designware_core    28672  1 i2c_designware_platform
i2c_designware_platform    16384  0
idma64                 20480  0
intel_lpss             16384  1 intel_lpss_pci
intel_lpss_pci         28672  0
intel_pmc_core         45056  0
intel_powerclamp       16384  0
intel_skl_int3472_discrete    16384  0
intel_skl_int3472_tps68470    20480  0
intel_tcc_cooling      16384  0
intel_vsec             16384  0
ipts                   16384  0
ithc                   36864  0
iwlmvm                360448  0
iwlwifi               311296  1 iwlmvm
ledtrig_audio          16384  1 snd_hda_codec_generic
libarc4                16384  1 mac80211
loop                   32768  42
mac80211              638976  1 iwlmvm
mc                     49152  4 v4l2_async,videodev,ov5693,ov13858
mei_pxp                16384  0
mousedev               24576  0
ov13858                24576  0
ov5693                 28672  0
pinctrl_tigerlake      32768  6
pmt_class              16384  1 pmt_telemetry
pmt_telemetry          16384  0
polyval_clmulni        16384  0
polyval_generic        16384  1 polyval_clmulni
rfcomm                 49152  16
sch_fq_codel           20480  1
serdev                 16384  1 surface_aggregator
serial_core            40960  3 8250,serial_mctrl_gpio,8250_base
serial_mctrl_gpio      16384  2 8250,8250_base
sg                     40960  0
snd                    94208  21 snd_hda_codec_generic,snd_seq,snd_seq_device,snd_hda_codec_hdmi,snd_hwdep,snd_seq_oss,snd_hda_intel,snd_hda_codec,snd_hda_codec_realtek,snd_sof,snd_timer,snd_compress,snd_soc_core,snd_pcm,snd_rawmidi
snd_compress           20480  1 snd_soc_core
snd_hda_codec         126976  5 snd_hda_codec_generic,snd_hda_codec_hdmi,snd_hda_intel,snd_hda_codec_realtek,snd_soc_hdac_hda
snd_hda_codec_generic    81920  1 snd_hda_codec_realtek
snd_hda_codec_hdmi     65536  1
snd_hda_codec_realtek   139264  1
snd_hda_core           69632  9 snd_hda_codec_generic,snd_hda_codec_hdmi,snd_hda_intel,snd_hda_ext_core,snd_hda_codec,snd_hda_codec_realtek,snd_sof_intel_hda_common,snd_soc_hdac_hda,snd_sof_intel_hda
snd_hda_ext_core       24576  3 snd_sof_intel_hda_common,snd_soc_hdac_hda,snd_sof_intel_hda
snd_hda_intel          36864  3
snd_hwdep              16384  1 snd_hda_codec
snd_intel_dspcfg       28672  3 snd_hda_intel,snd_sof,snd_sof_intel_hda_common
snd_intel_sdw_acpi     16384  2 snd_sof_intel_hda_common,snd_intel_dspcfg
snd_pcm               126976  11 snd_hda_codec_hdmi,snd_hda_intel,snd_hda_codec,soundwire_intel,snd_sof,snd_sof_intel_hda_common,snd_compress,snd_soc_core,snd_sof_utils,snd_hda_core,snd_pcm_dmaengine
snd_pcm_dmaengine      16384  1 snd_soc_core
snd_rawmidi            36864  1 snd_seq_midi
snd_seq                73728  6 snd_seq_midi,snd_seq_oss,snd_seq_midi_event,snd_seq_dummy
snd_seq_device         16384  4 snd_seq,snd_seq_midi,snd_seq_oss,snd_rawmidi
snd_seq_dummy          16384  0
snd_seq_midi           16384  0
snd_seq_midi_event     16384  2 snd_seq_midi,snd_seq_oss
snd_seq_oss            49152  0
snd_soc_acpi           16384  2 snd_soc_acpi_intel_match,snd_sof_intel_hda_common
snd_soc_acpi_intel_match    57344  2 snd_sof_intel_hda_common,snd_sof_pci_intel_tgl
snd_soc_core          229376  4 soundwire_intel,snd_sof,snd_sof_intel_hda_common,snd_soc_hdac_hda
snd_soc_hdac_hda       20480  1 snd_sof_intel_hda_common
snd_sof               184320  2 snd_sof_pci,snd_sof_intel_hda_common
snd_sof_intel_hda      16384  1 snd_sof_intel_hda_common
snd_sof_intel_hda_common   106496  1 snd_sof_pci_intel_tgl
snd_sof_pci            20480  2 snd_sof_intel_hda_common,snd_sof_pci_intel_tgl
snd_sof_pci_intel_tgl    16384  0
snd_sof_utils          16384  1 snd_sof
snd_sof_xtensa_dsp     16384  1 snd_sof_intel_hda_common
snd_timer              32768  2 snd_seq,snd_pcm
soc_button_array       16384  0
soundcore              16384  1 snd
soundwire_bus          77824  3 soundwire_intel,soundwire_generic_allocation,soundwire_cadence
soundwire_cadence      28672  1 soundwire_intel
soundwire_generic_allocation    16384  1 soundwire_intel
soundwire_intel        36864  1 snd_sof_intel_hda_common
squashfs               61440  21
surface_aggregator     65536  1 surface_aggregator_registry
surface_aggregator_registry    16384  0
surface_gpe            20480  0
surface_hotplug        16384  0
surfacepro3_button     16384  0
tls                   102400  0
tps68470_regulator     16384  0
usb_storage            65536  3
v4l2_async             20480  3 v4l2_fwnode,ov5693,ov13858
v4l2_fwnode            20480  2 ov5693,ov13858
videodev              217088  3 v4l2_async,ov5693,ov13858
virt_dma               16384  1 idma64
x86_pkg_temp_thermal    16384  0
```
I didn't enabled the Intel SPI drivers "Intel PCH/PCU SPI flash PCI driver" and "Intel PCH/PCU SPI flash platform driver" yet.

I don't see any possibility to attach my kernel config file in this discussion.

Can you tell me please, what I'm doing wrong?

**Reply by NP-chaonay** [reply]
> I've followed your instruction completely with the kernel source 6.0.7... Loading the module "ithc" generates dmesg output
> 
> ```
> [  447.828617] ithc: loading out-of-tree module taints kernel.
> [  447.829737] ithc 0000:00:10.6: enabling device (0000 -> 0002)
> [  448.049039] ithc 0000:00:10.6: config: e0000002 00000000 001fc0ff 00000010 00000014 43495424 09e34de8 0c37045e 00000001 e6eb47ab 1ae6d4ad 00000000 6e155668 00040302 3aad6885 c3ee4710
> [  448.049866] ithc 0000:00:10.6: registered device ithc
> [  448.050993] DMAR: DRHD: handling fault status reg 2
> [  448.051003] DMAR: [INTR-REMAP] Request device [01:05.0] fault index 0x40 [fault reason 0x26] Blocked an interrupt request due to source-id verification failure
> [  449.098841] ithc 0000:00:10.6: ithc_probe: hid_add_device failed with -110
> [  449.098856] ithc: probe of 0000:00:10.6 failed with error -110
> ```
> 
> I've configured almost relevant kernel drivers as modules. Maybe too much drivers...
> 
> ```
> 8250                   49152  1 8250_dw
> 8250_base              53248  2 8250,8250_dw
> 8250_dw                24576  0
> ac97_bus               16384  1 snd_soc_core
> acpi_pad               20480  0
> aesni_intel           393216  9
> af_alg                 28672  6 algif_hash,algif_skcipher
> algif_hash             16384  1
> algif_skcipher         16384  1
> atomisp_gmin_platform    28672  0
> binfmt_misc            16384  1
> bluetooth             634880  44 btrtl,btmtk,btintel,btbcm,bnep,btusb,rfcomm
> bnep                   20480  2
> btbcm                  24576  1 btusb
> btintel                36864  1 btusb
> btmtk                  16384  1 btusb
> btrtl                  24576  1 btusb
> btusb                  57344  0
> ccm                    20480  6
> cfg80211              430080  3 iwlmvm,iwlwifi,mac80211
> clk_tps68470           16384  0
> configfs               40960  1
> coretemp               16384  0
> crc32_pclmul           16384  0
> cryptd                 20480  3 crypto_simd,ghash_clmulni_intel
> crypto_simd            16384  1 aesni_intel
> ecc                    36864  1 ecdh_generic
> ecdh_generic           16384  2 bluetooth
> ghash_clmulni_intel    16384  0
> gpio_keys              20480  0
> i2c_designware_core    28672  1 i2c_designware_platform
> i2c_designware_platform    16384  0
> idma64                 20480  0
> intel_lpss             16384  1 intel_lpss_pci
> intel_lpss_pci         28672  0
> intel_pmc_core         45056  0
> intel_powerclamp       16384  0
> intel_skl_int3472_discrete    16384  0
> intel_skl_int3472_tps68470    20480  0
> intel_tcc_cooling      16384  0
> intel_vsec             16384  0
> ipts                   16384  0
> ithc                   36864  0
> iwlmvm                360448  0
> iwlwifi               311296  1 iwlmvm
> ledtrig_audio          16384  1 snd_hda_codec_generic
> libarc4                16384  1 mac80211
> loop                   32768  42
> mac80211              638976  1 iwlmvm
> mc                     49152  4 v4l2_async,videodev,ov5693,ov13858
> mei_pxp                16384  0
> mousedev               24576  0
> ov13858                24576  0
> ov5693                 28672  0
> pinctrl_tigerlake      32768  6
> pmt_class              16384  1 pmt_telemetry
> pmt_telemetry          16384  0
> polyval_clmulni        16384  0
> polyval_generic        16384  1 polyval_clmulni
> rfcomm                 49152  16
> sch_fq_codel           20480  1
> serdev                 16384  1 surface_aggregator
> serial_core            40960  3 8250,serial_mctrl_gpio,8250_base
> serial_mctrl_gpio      16384  2 8250,8250_base
> sg                     40960  0
> snd                    94208  21 snd_hda_codec_generic,snd_seq,snd_seq_device,snd_hda_codec_hdmi,snd_hwdep,snd_seq_oss,snd_hda_intel,snd_hda_codec,snd_hda_codec_realtek,snd_sof,snd_timer,snd_compress,snd_soc_core,snd_pcm,snd_rawmidi
> snd_compress           20480  1 snd_soc_core
> snd_hda_codec         126976  5 snd_hda_codec_generic,snd_hda_codec_hdmi,snd_hda_intel,snd_hda_codec_realtek,snd_soc_hdac_hda
> snd_hda_codec_generic    81920  1 snd_hda_codec_realtek
> snd_hda_codec_hdmi     65536  1
> snd_hda_codec_realtek   139264  1
> snd_hda_core           69632  9 snd_hda_codec_generic,snd_hda_codec_hdmi,snd_hda_intel,snd_hda_ext_core,snd_hda_codec,snd_hda_codec_realtek,snd_sof_intel_hda_common,snd_soc_hdac_hda,snd_sof_intel_hda
> snd_hda_ext_core       24576  3 snd_sof_intel_hda_common,snd_soc_hdac_hda,snd_sof_intel_hda
> snd_hda_intel          36864  3
> snd_hwdep              16384  1 snd_hda_codec
> snd_intel_dspcfg       28672  3 snd_hda_intel,snd_sof,snd_sof_intel_hda_common
> snd_intel_sdw_acpi     16384  2 snd_sof_intel_hda_common,snd_intel_dspcfg
> snd_pcm               126976  11 snd_hda_codec_hdmi,snd_hda_intel,snd_hda_codec,soundwire_intel,snd_sof,snd_sof_intel_hda_common,snd_compress,snd_soc_core,snd_sof_utils,snd_hda_core,snd_pcm_dmaengine
> snd_pcm_dmaengine      16384  1 snd_soc_core
> snd_rawmidi            36864  1 snd_seq_midi
> snd_seq                73728  6 snd_seq_midi,snd_seq_oss,snd_seq_midi_event,snd_seq_dummy
> snd_seq_device         16384  4 snd_seq,snd_seq_midi,snd_seq_oss,snd_rawmidi
> snd_seq_dummy          16384  0
> snd_seq_midi           16384  0
> snd_seq_midi_event     16384  2 snd_seq_midi,snd_seq_oss
> snd_seq_oss            49152  0
> snd_soc_acpi           16384  2 snd_soc_acpi_intel_match,snd_sof_intel_hda_common
> snd_soc_acpi_intel_match    57344  2 snd_sof_intel_hda_common,snd_sof_pci_intel_tgl
> snd_soc_core          229376  4 soundwire_intel,snd_sof,snd_sof_intel_hda_common,snd_soc_hdac_hda
> snd_soc_hdac_hda       20480  1 snd_sof_intel_hda_common
> snd_sof               184320  2 snd_sof_pci,snd_sof_intel_hda_common
> snd_sof_intel_hda      16384  1 snd_sof_intel_hda_common
> snd_sof_intel_hda_common   106496  1 snd_sof_pci_intel_tgl
> snd_sof_pci            20480  2 snd_sof_intel_hda_common,snd_sof_pci_intel_tgl
> snd_sof_pci_intel_tgl    16384  0
> snd_sof_utils          16384  1 snd_sof
> snd_sof_xtensa_dsp     16384  1 snd_sof_intel_hda_common
> snd_timer              32768  2 snd_seq,snd_pcm
> soc_button_array       16384  0
> soundcore              16384  1 snd
> soundwire_bus          77824  3 soundwire_intel,soundwire_generic_allocation,soundwire_cadence
> soundwire_cadence      28672  1 soundwire_intel
> soundwire_generic_allocation    16384  1 soundwire_intel
> soundwire_intel        36864  1 snd_sof_intel_hda_common
> squashfs               61440  21
> surface_aggregator     65536  1 surface_aggregator_registry
> surface_aggregator_registry    16384  0
> surface_gpe            20480  0
> surface_hotplug        16384  0
> surfacepro3_button     16384  0
> tls                   102400  0
> tps68470_regulator     16384  0
> usb_storage            65536  3
> v4l2_async             20480  3 v4l2_fwnode,ov5693,ov13858
> v4l2_fwnode            20480  2 ov5693,ov13858
> videodev              217088  3 v4l2_async,ov5693,ov13858
> virt_dma               16384  1 idma64
> x86_pkg_temp_thermal    16384  0
> ```
> 
> I didn't enabled the Intel SPI drivers "Intel PCH/PCU SPI flash PCI driver" and "Intel PCH/PCU SPI flash platform driver" yet.
> 
> I don't see any possibility to attach my kernel config file in this discussion.
> 
> Can you tell me please, what I'm doing wrong?

1) 

> DMAR: [INTR-REMAP] Request device [01:05.0] fault index 0x40 [fault reason 0x26] Blocked an interrupt request due to source-id verification failure

see this ```If you get an error in dmesg saying "Blocked an interrupt request due to source-id verification failure",``` at https://github.com/quo/ithc-linux

2) 

> Maybe too much drivers...

i think it is okey, let see my output
[1.txt](https://github.com/linux-surface/linux-surface/files/9953517/1.txt)

**Comment by mickpf** [comment]
Hello,

I followed this instruction(s), get much better dmesg output
```
    [    5.012785] ithc: loading out-of-tree module taints kernel.
    [    5.019166] ithc 0000:00:10.6: enabling device (0000 -> 0002)
    [    5.237260] ithc 0000:00:10.6: config: e0000002 00000000 001fc0ff 00000010 00000014 43495424 09e3dde8 0c37045e 00000001 e4eb43ab 1ae6c4ad 00000000 6e15526c 00040302 7aafa885 c3ee4710
    [    5.237776] ithc 0000:00:10.6: registered device ithc
    [    5.241209] input: Intel Touch Host Controller as /devices/pci0000:00/0000:00:10.6/0001:045E:0C37.0004/input/input9
    [    5.241290] input: Intel Touch Host Controller as /devices/pci0000:00/0000:00:10.6/0001:045E:0C37.0004/input/input10
    [    5.241349] input: Intel Touch Host Controller as /devices/pci0000:00/0000:00:10.6/0001:045E:0C37.0004/input/input11
    [    5.241404] input: Intel Touch Host Controller Touchscreen as /devices/pci0000:00/0000:00:10.6/0001:045E:0C37.0004/input/input12
    [    5.241476] input: Intel Touch Host Controller as /devices/pci0000:00/0000:00:10.6/0001:045E:0C37.0004/input/input13
    [    5.241526] input: Intel Touch Host Controller as /devices/pci0000:00/0000:00:10.6/0001:045E:0C37.0004/input/input14
    [    5.241586] input: Intel Touch Host Controller Stylus as /devices/pci0000:00/0000:00:10.6/0001:045E:0C37.0004/input/input15
    [    5.241658] input: Intel Touch Host Controller as /devices/pci0000:00/0000:00:10.6/0001:045E:0C37.0004/input/input16
    [    5.375221] hid-generic 0001:045E:0C37.0004: input,hidraw3: <UNKNOWN> HID v1.00 Device [Intel Touch Host Controller] on pci-0000:00:10.6/ithc
```
I've installed the daemon "iptsd"
```
    [ 1084.877639] input: IPTS Touch as /devices/virtual/input/input31
    [ 1084.877703] input: IPTS Stylus as /devices/virtual/input/input32
```
BUT THE KEYBOARD AND THE TOUCHPAD DO NOT WORK YET!!!
Sorry, I'm very angry about this...

Is it necessary to boot without the external keyboard?

Can you help me please?

Thanks in advance,
Michael

**Reply by qzed** [reply]
Touch/ITHC isn't required for the keyboard/typecover, it's only for touch. Based on the list of loaded modules you've posted above, you seem to be missing `surface_hid` and `surface_hid_core`. Also `surface_battery` and `surface_charger`, so I assume battery status also doesn't show up?

**Comment by mickpf** [comment]
Hello Max,

could we communicate in German, it's much easier to me!
Well, I've loaded these modules manually, but as you imagine - without any output in dmesg and without any success!

Ich bin frustriert, weil ich gehofft habe, das Surface auf Reisen mitzunehmen UND dabei vorwiegend Linux nutzen zu können. Und nicht nur ständig auf das Windows angewiesen zu sein! ich habe einige Entwicklungstools, die ich ausschließlich unter Linux nutze...

**Reply by qzed** [reply]
Könntest du das `diagnostics.sh` skript wie [hier](https://github.com/linux-surface/linux-surface/wiki/Testers-and-Contributing#support-for-new-devices) erwähnt ausführen und das ergebnis hochladen (falls das nicht hier geht einfach einen neuen Issue erstellen)?

**Reply by StollD** [reply]
Eine Sache die beim Kompilieren von eigenen Kerneln immer wieder Probleme macht sind die Pinctrl Treiber. Überprüf mal, ob eines dieser Module in deinem Kernel aktiviert ist:

```
CONFIG_PINCTRL_BAYTRAIL=y
CONFIG_PINCTRL_CHERRYVIEW=y
CONFIG_PINCTRL_LYNXPOINT=m
CONFIG_PINCTRL_INTEL=y
CONFIG_PINCTRL_ALDERLAKE=m
CONFIG_PINCTRL_BROXTON=m
CONFIG_PINCTRL_CANNONLAKE=m
CONFIG_PINCTRL_CEDARFORK=m
CONFIG_PINCTRL_DENVERTON=m
CONFIG_PINCTRL_ELKHARTLAKE=m
CONFIG_PINCTRL_EMMITSBURG=m
CONFIG_PINCTRL_GEMINILAKE=m
CONFIG_PINCTRL_ICELAKE=m
CONFIG_PINCTRL_JASPERLAKE=m
CONFIG_PINCTRL_LAKEFIELD=m
CONFIG_PINCTRL_LEWISBURG=m
CONFIG_PINCTRL_METEORLAKE=m
CONFIG_PINCTRL_SUNRISEPOINT=m
CONFIG_PINCTRL_TIGERLAKE=m
```

`CONFIG_PINCTRL_TIGERLAKE` müsste das sein das du für das SP8 brauchst, aber ich bin bei den Lakes von Intel nicht so auf dem Laufenden.

**Reply by mickpf** [reply]
Danke für den Hinweis...
Nun, dieser Benutzerfehler ist mir sehr wohlbekannt, deshalb habe ich es auch richtig gemacht (s. Anhang weiter oben oder unten), das Modul "pinctrl_tigerlake" ist auch geladen...

**Comment by mickpf** [comment]
Ich sehe eine Möglichkeit für das Hochladen von Dateien, die ich bisher übersehen habe... Ich glaube, ich brauche eine Brille für Augenlose... :-)
Im Anhang findest du die Ausgabe des Befehls "diagnostics.sh full"
[diagnostics.tar.gz](https://github.com/linux-surface/linux-surface/files/9963920/diagnostics.tar.gz)

**Reply by qzed** [reply]
Ah, ich glaube ich habe den Fehler gefunden: Du brauchst
```
CONFIG_SERIAL_DEV_BUS=y
SERIAL_DEV_CTRL_TTYPORT=y
```
Im moment hast du nur
```
CONFIG_SERIAL_DEV_BUS=m
```

**Comment by mickpf** [comment]
Ich probiere es... Es wird etwas dauern

Danke für den Hinweis!

**Comment by mickpf** [comment]
Ja, das war's!!!

Jetzt funktionieren die Tastatur, das Touchpad und die Batterie-Anzeige!

Vielen Dank für die Hilfe!
